frontend/src/stimulus/controllers/pattern-autocompleter.controller.ts

```diff
@@ -0,0 +1,305 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * See COPYRIGHT and LICENSE files for more details.
+ * ++
+ */
+
+import { Controller } from '@hotwired/stimulus';
+
+export default class PatternAutocompleterController extends Controller {
+  static targets = [
+    'tokenTemplate',
+    'content',
+    'formInput',
+    'suggestions',
+  ];
+
+  declare readonly tokenTemplateTarget:HTMLTemplateElement;
+  declare readonly contentTarget:HTMLElement;
+  declare readonly formInputTarget:HTMLInputElement;
+  declare readonly suggestionsTarget:HTMLElement;
+
+  static values = { patternInitial: String };
+  declare patternInitialValue:string;
+
+  // internal state
+  currentRange:Range|undefined = undefined;
+  selectedSuggestion:{ element:HTMLElement|null, index:number } = { element: null, index: 0 };
+
+  connect() {
+    this.contentTarget.innerHTML = this.toHtml(this.patternInitialValue) || ' ';
+  }
+
+  // Input field events
+  input_keydown(event:KeyboardEvent) {
+    // insert the selected suggestion
+    if (event.key === 'Enter') {
+      // prevent entering new line characters
+      event.preventDefault();
+
+      const selectedItem = this.suggestionsTarget.querySelector('.selected') as HTMLElement;
+      if (selectedItem) {
+        this.insertToken(this.createToken(selectedItem.dataset.value!));
+        this.clearSuggestionsFilter();
+      }
+    }
+
+    // move up and down the suggestions selection
+    if (event.key === 'ArrowUp') {
+      event.preventDefault();
+      this.selectSuggestionAt(this.selectedSuggestion.index - 1);
+    }
+    if (event.key === 'ArrowDown') {
+      event.preventDefault();
+      this.selectSuggestionAt(this.selectedSuggestion.index + 1);
+    }
+
+    // close the suggestions
+    if (event.key === 'Escape') {
+      this.hide(this.suggestionsTarget);
+    }
+
+    // update cursor
+    this.setRange();
+  }
+
+  input_change() {
+    // browsers insert a `<br>` tag on empty contenteditable elements so we need to cleanup
+    if (this.contentTarget.innerHTML === '<br>') {
+      this.contentTarget.innerHTML = ' ';
+    }
+
+    this.ensureSpacesAround();
+
+    // show suggestions for the current word
+    const word = this.currentWord();
+    if (word && word.length > 0) {
+      this.filterSuggestions(word);
+      this.selectSuggestionAt(0);
+      this.show(this.suggestionsTarget);
+    } else {
+      this.clearSuggestionsFilter();
+    }
+
+    // update cursor
+    this.setRange();
+  }
+
+  input_mouseup() {
+    this.setRange();
+  }
+
+  input_focus() {
+    this.setRange();
+  }
+
+  input_blur() {
```

Comment:
I think blur is also a good moment to close the auto-completion. I noticed that I automatically (i.e. through muscle memory) clicked next to the input when I wanted the dropdown to close and then I was surprised that it remained open.

Reply:
This was the case at some point. I don't remember why I removed it, but I'll add again.